Feature/private submission

Vkale1 opened the pull request:
This is not to be merged. It is just a review for the first part change of the code so you don't get a massive batch of changes to review at once. This change solely tackles the API endpoints. Any comments on cleaner coding would be good. I will tackle the exceptions and tests in a later review.

Logic:
--private is an input parameter 
Public - portal API search endpoints with no authentication.
Private: 
- Webin reports API. JSON if required metadata is within it otherwise XML.
- All fields are renamed to match the fields that the public API returns.
- Lat/Long formats are different.

Ge94 reviewed (approved):
Massive work for a ridiculous problem... Thank you lots V.
I think that as long as it works let's keep it like this until we know for certain that we need to reimplement it with the new webin-cli (unless we recycle this code for the assembly_uploader?). But for the future I'd suggest:

- Reformat all URLs as constants (also the one below, in the genome submission part)
- Write functions for post/get requests and data parsing/processing
- This could include writing a set of constants for exceptions

But again... I'd leave this for now

Vkale1 commented:
New changes!
- Separated calls to the private and public APIs
- Add --private option. Try private first in this case and public afterwards
- Reformatted to remove camelCase for variables and functions
- Moved around functions which fit in the class to remove redundancy. Main function is simplified.
- Reformatted so that genome uploader can be useable as a python library.
- Removed command line credentials. Should be in a config, .env or exported new before. Credentials are handled in ena.py now. Added defaults and requirements to other arguments.
- Unit tests for all ena calls and one exception test added.
- Standard pre-commit formatting checks applied to all files.

The end-to-end test doesn't work now.. it's failing at the submission of the samples with a Forbidden error. I suspect it's a credential thing but I'm not sure how you got around this before?

Ge94 commented on genomeuploader/constants.py:
```diff
@@ -14,14 +14,8 @@
-    "Many fragments with little to no review of assembly other than reporting "
-    "of standard assembly statistics."
-)
+HQ = "Multiple fragments where gaps span repetitive regions. Presence of the " "23S, 16S, and 5S rRNA genes and at least 18 tRNAs."
```
```suggestion
HQ = "Multiple fragments where gaps span repetitive regions. Presence of the 23S, 16S, and 5S rRNA genes and at least 18 tRNAs."
```

Ge94 commented on genomeuploader/constants.py:
```diff
@@ -14,14 +14,8 @@
-    "of standard assembly statistics."
-)
+HQ = "Multiple fragments where gaps span repetitive regions. Presence of the " "23S, 16S, and 5S rRNA genes and at least 18 tRNAs."
+MQ = "Many fragments with little to no review of assembly other than reporting " "of standard assembly statistics."
```
```suggestion
MQ = "Many fragments with little to no review of assembly other than reporting of standard assembly statistics."
```

Ge94 commented on genomeuploader/ena.py:
```diff
@@ -32,265 +36,396 @@ class NoDataException(ValueError):
-        url = "https://www.ebi.ac.uk/ena/portal/api/search"
-        auth = (webin, password)
-        default_connection_headers = {
+
```
```suggestion
```

Ge94 commented on genomeuploader/ena.py:
```diff
@@ -32,265 +36,396 @@ class NoDataException(ValueError):
+        self.acc_type = parse_accession(accession)
+        username, password = configure_credentials()
+        if username is None or password is None:
+            logging.error("ENA_WEBIN and ENA_WEBIN_PASSWORD are not set")
```
Is this stopping the execution or just logging?

Ge94 commented on genomeuploader/ena.py:
```diff
@@ -32,265 +36,396 @@ class NoDataException(ValueError):
+        self.browser_url = "https://www.ebi.ac.uk/ena/browser/api/xml"
+        self.accession = accession
+        self.acc_type = parse_accession(accession)
+        username, password = configure_credentials()
+        if username is None or password is None:
+            logging.error("ENA_WEBIN and ENA_WEBIN_PASSWORD are not set")
+        if username and password:
+            self.auth = (username, password)
+        else:
+            self.auth = None
```
We always need the credentials for submission - we might remove them when querying ENA publicly, but we'll need to raise an error if no credentials are available.

Ge94 commented on genomeuploader/genome_upload.py:
```diff
@@ -712,230 +534,453 @@ def write_genomes_xml(genomes, xml_path, genomeType, centreName, tpa):
+            accessionsgen = accessionsgen.replace(".tsv", "_test.tsv")
+
+        accessions_file = os.path.join(self.upload_dir, accessionsgen)
+        save = False
+        write_mode = "a"
+        if os.path.exists(accessions_file):
+            if not self.live:
+                save = True
+                if self.force:
+                    write_mode = "w"
+            if not save:
+                logger.info("Genome samples already registered, reading ERS accessions...")
+                alias_to_new_sample_accession = get_accessions(accessions_file)
+        else:
+            save = True
```
Note to self: this affects line 956 too

Vkale1 replied:
No logic has changed AFAIK. The only change was moving common arguments into a class. This is exactly this snippet but within the class 
https://github.com/EBI-Metagenomics/genome_uploader/blob/0a24eca14384e5828fc455ff1212fcd7f3936988/genomeuploader/genome_upload.py#L813C1-L825C20

MGS-sails commented on genomeuploader/ena.py:
```diff
@@ -32,265 +36,396 @@ class NoDataException(ValueError):
+        self.accession = accession
+        self.acc_type = parse_accession(accession)
+        username, password = configure_credentials()
+        if username is None or password is None:
```
I think we could make this more declarative by separating out the auth logic. 
something like this for instance:

MGS-sails reviewed (commented):
Thanks @Ge94 .
I had quite a few suggestions. 
Thought it might help if I just committed my changes. 
I've made 2 new commits that have refactored the code as such:

1. Separation of concerns: Moved constant values into a config.py file, moved EnaSubmit into a separate file. Created a dedicated exceptions.py file. 
2. More error handling using dedicated custom error types.
3. Extracted repetitive coding patterns into smaller functions:

Refactor Highlights:
Replaced get_data_or_handle_error with get_data_or_raise
Introduced clearer, exception-based control flow that raises specific errors (EnaEmptyResponseError, EnaParseError) instead of returning None on failure. This avoids silent failures and improves debuggability.

Updated _fetch_ena_data
Adjusted to catch and log parse and response-related errors raised by get_data_or_raise, returning None explicitly when an error occurs, while keeping the reformatter logic intact.

Improved structure and readability
Parsing logic is centralized using a dictionary of lambdas (parsers), eliminating repetitive if statements and making the function more declarative and maintainable.

Custom exception classes added
Defined specific error types to capture different failure scenarios more explicitly. While some may appear minimal now, they provide a clear taxonomy for handling and extending future error management.

Vkale1 commented on genomeuploader/ena.py:
```diff
@@ -32,265 +36,396 @@ class NoDataException(ValueError):
+        return "analysis_accession"
+    else:
+        logging.error(f"{accession} is not a valid accession")
+        sys.exit()
```
Is config.py similar to constants.py here? Could I combine both?

Vkale1 commented on genomeuploader/genome_upload.py:
```diff
@@ -14,34 +14,72 @@
+
+    # Users can provide their credentials in environment variables or using a config file
+    parser.add_argument("--centre_name", required=True, help="Name of the centre uploading genomes")
+    parser.add_argument(
```
I think the point of removing this is that I don't want to be reading credentials from the command line. They should be in the env file or loaded into the environment. I will change the readme accordingly but args.webin and args.password are no longer used anywhere in the scripts.

Vkale1 commented on genomeuploader/genome_upload.py:
```diff
@@ -14,34 +14,72 @@
+
+    parser.add_argument("--out", type=str, help="Output folder. Default: working directory")
+    parser.add_argument("--force", action="store_true", required=False, default=False, help="Forces reset of sample xml's backups")
+    parser.add_argument(
+        "--live",
+        action="store_true",
+        required=False,
+        default=False,
+        help="Uploads on ENA. Omitting this " + "option allows to validate samples beforehand",
+    )
```
Yeah so.. I agree with you but recommit-checks don't I'm afraid. It keeps undoing it sorry

Ge94 reviewed (approved):
Nothing to say, it looks great! Thank you both for your time spent on this. Happy to give an extra look to this once rebased with main

Ge94 commented on genomeuploader/ena.py:
```diff
@@ -16,281 +15,361 @@
+import requests
+from dotenv import load_dotenv
+from requests.exceptions import ConnectionError, HTTPError, RequestException, Timeout
+from .config import ACCESSION_MAP, USER_ENV_FILE_PATH, RUN_KEY_RENAME_MAP, SAMPLE_ATTRIBUTE_MAP, RETRY_COUNT, ENA_SUBMISSION_URL, ENA_SEARCH_URL, ENA_BROWSER_URL,
```
Old leftover comment

MGS-sails reviewed (approved):
Thanks @Vkale1 generally looks good. 
I added a minor comment on using the crednetials_manager.
